lib/publiccloud/instance.pm

```diff
@@ -350,55 +352,47 @@ sub wait_for_ssh {
     $args{username} //= $self->username();
     $args{public_ip} //= $self->public_ip();
     my $start_time = time();
-    my $check_port = 1;
     my $sleep_period = $args{ignore_wrong_pubkey} ? 20 : 1;
-
-    # Looping until reaching timeout or passing two conditions :
-    # - SSH port 22 is reachable
-    # - journalctl got message about reaching one of certain targets
-    while ((my $duration = time() - $start_time) < $args{timeout}) {
-        if ($check_port) {
-            $check_port = 0 if (script_run('nc -vz -w 1 ' . $self->{public_ip} . ' 22', quiet => 1) == 0);
-        }
-        else {
-            # On boottime test we do hard reboot which may change the instance address
-            script_run("ssh-keyscan $args{public_ip} | tee -a ~/.ssh/known_hosts") if (get_var('PUBLIC_CLOUD_CHECK_BOOT_TIME'));
-
-            my $output = $self->run_ssh_command(
-                cmd => 'sudo journalctl -b | grep -E "Reached target (Cloud-init|Default|Main User Target)"',
-                proceed_on_failure => 1,
-                username => $args{username});
-            if ($output =~ m/Reached target.*/) {
-                return $duration;
-            }
-            elsif ($output =~ m/Permission denied \(publickey\).*/) {
-                die "ssh permission denied (pubkey)" unless $args{ignore_wrong_pubkey};
-            }
-        }
-        sleep $sleep_period;
+    my $retry = ($args{timeout} > $sleep_period) ? int($args{timeout} / $sleep_period) : 1;
+    my $status = 0;
+    my $out = '#';
+    my $instance_msg = sprintf("instance %s with public IP:%s within %d seconds", $self->{instance_id}, $self->{public_ip}, $args{timeout});
+    my $duration;
+    # Looping until SSH port 22 is reachable or timeout:
+    script_retry('nc -vz -w 1 ' . $self->{public_ip} . ' 22', timeout => $args{timeout}, delay => $sleep_period, retry => $retry, fail_msg => "Unable to reach SSH port of " . $instance_msg);
+    # ssh OK:
+    $duration = time() - $start_time;    # time after ssh ok
+
+    # help in boottime tests,when hard reboot could change the instance address:
+    $self->ssh_script_run("ssh-keyscan $args{public_ip} | tee -a ~/.ssh/known_hosts") if (get_var('PUBLIC_CLOUD_PERF_COLLECT') or get_var('PUBLIC_CLOUD_CHECK_BOOT_TIME'));
+
+    # check remote system is running, looping by --wait option:
+    $out = $self->ssh_script_output(cmd => 'sudo systemctl is-system-running --wait', timeout => ($args{timeout} - $duration), proceed_on_failure => 1, username => $args{username});
```

Comment:
`timeout => ($args{timeout} - $duration)` - can we get negative value here ? and what would be the behavior of `ssh_script_output`

Reply:
Note: the `timeout = 0` usually in our functions **deactivates** the timeout